server/tls.go

```diff
@@ -0,0 +1,58 @@
+// This code is derived from https://github.com/coredns/coredns
+package server
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+)
+
+// NewTLSConfig returns a TLS config that includes a certificate
+// Use for Server TLS config or when using a client certificate
+// If caPath is empty, system CAs will be used
+func NewTLSConfig(certPath, keyPath, caPath string) (*tls.Config, error) {
+	cert, err := tls.LoadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	roots, err := loadRoots(caPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      roots,
```

Comment:
For server certs, this should be ClientCAs, not RootCAs. Yes, this is a bug in the CoreDNS TLS code, I believe. I will probably need to fix that to get client-based cert working there.

Reply:
Also make sure this work in combination with a hand-created tls.Listener - I think it should if you use the same tlsConfig for both. But I seem to recall when I was playing with it that I couldn't get client-certs to work with using my own tls Listener.